File: state_test.go
package main

import (
	"testing"
	"time"

	"github.com/stretchr/testify/suite"
)

// --- Item State Machine Suite ---

type StateMachineSuite struct {
	suite.Suite
	sm *StateMachine
}

func (s *StateMachineSuite) SetupTest() {
	s.sm = NewStateMachine()
}

func (s *StateMachineSuite) transition(kind TransitionKind, from ItemStatus, evidence, reason string) (ItemStatus, error) {
	return s.sm.Apply(Transition{
		Kind:      kind,
		From:      from,
		Evidence:  evidence,
		Reason:    reason,
		Actor:     "tester",
		Timestamp: time.Now(),
	})
}

func (s *StateMachineSuite) TestVerifyFromUnverified() {
	got, err := s.transition(TransitionVerify, StatusUnverified, "test passed", "")
	s.Require().NoError(err)
	s.Equal(StatusProven, got)
}

func (s *StateMachineSuite) TestSuspectFromProven() {
	got, err := s.transition(TransitionSuspect, StatusProven, "", "upstream changed")
	s.Require().NoError(err)
	s.Equal(StatusSuspect, got)
}

func (s *StateMachineSuite) TestReVerifyFromSuspect() {
	got, err := s.transition(TransitionReVerify, StatusSuspect, "re-tested", "still valid")
	s.Require().NoError(err)
	s.Equal(StatusProven, got)
}

func (s *StateMachineSuite) TestBreakFromSuspect() {
	got, err := s.transition(TransitionBreak, StatusSuspect, "", "confirmed broken")
	s.Require().NoError(err)
	s.Equal(StatusBroke, got)
}

func (s *StateMachineSuite) TestFixFromBroke() {
	got, err := s.transition(TransitionFix, StatusBroke, "fixed and tested", "bug resolved")
	s.Require().NoError(err)
	s.Equal(StatusProven, got)
}

func (s *StateMachineSuite) TestInvalidTransitions() {
	tests := []struct {
		name string
		from ItemStatus
		kind TransitionKind
	}{
		{"unverified cannot break", StatusUnverified, TransitionBreak},
		{"proven cannot verify again", StatusProven, TransitionVerify},
		{"broke cannot become suspect", StatusBroke, TransitionSuspect},
		{"unverified cannot re-verify", StatusUnverified, TransitionReVerify},
		{"proven cannot fix", StatusProven, TransitionFix},
	}
	for _, tt := range tests {
		s.Run(tt.name, func() {
			_, err := s.transition(tt.kind, tt.from, "evidence", "reason")
			s.Error(err)
		})
	}
}

func (s *StateMachineSuite) TestVerifyRequiresEvidence() {
	_, err := s.transition(TransitionVerify, StatusUnverified, "", "")
	s.Error(err)
	s.Contains(err.Error(), "requires evidence")
}

func (s *StateMachineSuite) TestReVerifyRequiresEvidenceAndReason() {
	_, err := s.transition(TransitionReVerify, StatusSuspect, "", "reason")
	s.Error(err)
	s.Contains(err.Error(), "requires evidence")

	_, err = s.transition(TransitionReVerify, StatusSuspect, "evidence", "")
	s.Error(err)
	s.Contains(err.Error(), "requires a reason")
}

func (s *StateMachineSuite) TestRequiresActor() {
	_, err := s.sm.Apply(Transition{
		Kind:      TransitionVerify,
		From:      StatusUnverified,
		Evidence:  "proof",
		Actor:     "",
		Timestamp: time.Now(),
	})
	s.Error(err)
	s.Contains(err.Error(), "requires an actor")
}

func (s *StateMachineSuite) TestAvailableTransitions() {
	rules := s.sm.AvailableTransitions(StatusSuspect)
	s.Len(rules, 2) // re_verify and break

	kinds := make(map[TransitionKind]bool)
	for _, r := range rules {
		kinds[r.Kind] = true
	}
	s.True(kinds[TransitionReVerify])
	s.True(kinds[TransitionBreak])
}

func (s *StateMachineSuite) TestCanTransition() {
	s.True(s.sm.CanTransition(StatusUnverified, TransitionVerify))
	s.False(s.sm.CanTransition(StatusUnverified, TransitionBreak))
	s.True(s.sm.CanTransition(StatusProven, TransitionSuspect))
	s.False(s.sm.CanTransition(StatusProven, TransitionVerify))
}

func TestStateMachineSuite(t *testing.T) {
	suite.Run(t, new(StateMachineSuite))
}

// --- CR State Machine Suite ---

type CRStateMachineSuite struct {
	suite.Suite
	sm *CRStateMachine
}

func (s *CRStateMachineSuite) SetupTest() {
	s.sm = NewCRStateMachine()
}

func (s *CRStateMachineSuite) TestFullApprovalLifecycle() {
	steps := []struct {
		from CRStatus
		kind CRTransitionKind
		want CRStatus
	}{
		{CRDraft, CRSubmit, CRProposed},
		{CRProposed, CROpen, CRVoting},
		{CRVoting, CRApprove, CRApproved},
		{CRApproved, CRApplyT, CRApplied},
		{CRApplied, CRArchive, CRArchived},
	}

	current := CRDraft
	for _, step := range steps {
		s.Require().Equal(step.from, current)
		got, err := s.sm.Apply(current, step.kind)
		s.Require().NoError(err, "step %s->%s", step.from, step.want)
		s.Equal(step.want, got)
		current = got
	}
}

func (s *CRStateMachineSuite) TestRejectPath() {
	got, err := s.sm.Apply(CRVoting, CRReject)
	s.Require().NoError(err)
	s.Equal(CRRejected, got)

	got, err = s.sm.Apply(CRRejected, CRArchive)
	s.Require().NoError(err)
	s.Equal(CRArchived, got)
}

func (s *CRStateMachineSuite) TestInvalidTransitions() {
	tests := []struct {
		name string
		from CRStatus
		kind CRTransitionKind
	}{
		{"draft cannot approve", CRDraft, CRApprove},
		{"draft cannot reject", CRDraft, CRReject},
		{"proposed cannot approve", CRProposed, CRApprove},
		{"approved cannot reject", CRApproved, CRReject},
		{"archived cannot do anything", CRArchived, CRSubmit},
	}
	for _, tt := range tests {
		s.Run(tt.name, func() {
			_, err := s.sm.Apply(tt.from, tt.kind)
			s.Error(err)
		})
	}
}

func TestCRStateMachineSuite(t *testing.T) {
	suite.Run(t, new(CRStateMachineSuite))
}

File: packages/node/test/state.test.ts
import { describe, expect, test } from "bun:test";
import type { ItemState, TransitionKind } from "@inv/shared";
import {
  StateMachine,
  CRStateMachine,
  type CRStatus,
  type CRTransitionKind,
} from "../src/state";

describe("StateMachine", () => {
  const sm = new StateMachine();

  describe("valid transitions", () => {
    test("verify: unverified -> proven", () => {
      expect(sm.apply("unverified", "verify")).toBe("proven");
    });

    test("suspect: proven -> suspect", () => {
      expect(sm.apply("proven", "suspect")).toBe("suspect");
    });

    test("re_verify: suspect -> proven", () => {
      expect(sm.apply("suspect", "re_verify")).toBe("proven");
    });

    test("break: suspect -> broke", () => {
      expect(sm.apply("suspect", "break")).toBe("broke");
    });

    test("fix: broke -> proven", () => {
      expect(sm.apply("broke", "fix")).toBe("proven");
    });
  });

  describe("invalid transitions throw", () => {
    test("cannot verify from proven", () => {
      expect(() => sm.apply("proven", "verify")).toThrow();
    });

    test("cannot suspect from unverified", () => {
      expect(() => sm.apply("unverified", "suspect")).toThrow();
    });

    test("cannot re_verify from unverified", () => {
      expect(() => sm.apply("unverified", "re_verify")).toThrow();
    });

    test("cannot break from proven", () => {
      expect(() => sm.apply("proven", "break")).toThrow();
    });

    test("cannot fix from proven", () => {
      expect(() => sm.apply("proven", "fix")).toThrow();
    });

    test("cannot break from unverified", () => {
      expect(() => sm.apply("unverified", "break")).toThrow();
    });

    test("cannot fix from suspect", () => {
      expect(() => sm.apply("suspect", "fix")).toThrow();
    });

    test("cannot verify from broke", () => {
      expect(() => sm.apply("broke", "verify")).toThrow();
    });
  });

  describe("canTransition", () => {
    test("returns true for valid transition", () => {
      expect(sm.canTransition("unverified", "verify")).toBe(true);
    });

    test("returns true for suspect -> re_verify", () => {
      expect(sm.canTransition("suspect", "re_verify")).toBe(true);
    });

    test("returns true for suspect -> break", () => {
      expect(sm.canTransition("suspect", "break")).toBe(true);
    });

    test("returns true for broke -> fix", () => {
      expect(sm.canTransition("broke", "fix")).toBe(true);
    });

    test("returns false for invalid transition", () => {
      expect(sm.canTransition("proven", "verify")).toBe(false);
    });

    test("returns false for unverified -> fix", () => {
      expect(sm.canTransition("unverified", "fix")).toBe(false);
    });

    test("returns false for broke -> suspect", () => {
      expect(sm.canTransition("broke", "suspect")).toBe(false);
    });
  });

  describe("availableTransitions", () => {
    test("unverified has only verify", () => {
      expect(sm.availableTransitions("unverified")).toEqual(["verify"]);
    });

    test("proven has only suspect", () => {
      expect(sm.availableTransitions("proven")).toEqual(["suspect"]);
    });

    test("suspect has re_verify and break", () => {
      const transitions = sm.availableTransitions("suspect");
      expect(transitions).toContain("re_verify");
      expect(transitions).toContain("break");
      expect(transitions).toHaveLength(2);
    });

    test("broke has only fix", () => {
      expect(sm.availableTransitions("broke")).toEqual(["fix"]);
    });
  });
});

describe("CRStateMachine", () => {
  const cr = new CRStateMachine();

  describe("valid transitions", () => {
    test("submit: draft -> proposed", () => {
      expect(cr.apply("draft", "submit")).toBe("proposed");
    });

    test("open_voting: proposed -> voting", () => {
      expect(cr.apply("proposed", "open_voting")).toBe("voting");
    });

    test("approve: voting -> approved", () => {
      expect(cr.apply("voting", "approve")).toBe("approved");
    });

    test("reject: voting -> rejected", () => {
      expect(cr.apply("voting", "reject")).toBe("rejected");
    });

    test("apply: approved -> applied", () => {
      expect(cr.apply("approved", "apply")).toBe("applied");
    });

    test("archive from applied: applied -> archived", () => {
      expect(cr.apply("applied", "archive")).toBe("archived");
    });

    test("archive from rejected: rejected -> archived", () => {
      expect(cr.apply("rejected", "archive")).toBe("archived");
    });
  });

  describe("invalid transitions throw", () => {
    test("cannot submit from proposed", () => {
      expect(() => cr.apply("proposed", "submit")).toThrow();
    });

    test("cannot open_voting from draft", () => {
      expect(() => cr.apply("draft", "open_voting")).toThrow();
    });

    test("cannot approve from proposed", () => {
      expect(() => cr.apply("proposed", "approve")).toThrow();
    });

    test("cannot reject from proposed", () => {
      expect(() => cr.apply("proposed", "reject")).toThrow();
    });

    test("cannot apply from voting", () => {
      expect(() => cr.apply("voting", "apply")).toThrow();
    });

    test("cannot archive from draft", () => {
      expect(() => cr.apply("draft", "archive")).toThrow();
    });

    test("cannot archive from proposed", () => {
      expect(() => cr.apply("proposed", "archive")).toThrow();
    });

    test("cannot archive from voting", () => {
      expect(() => cr.apply("voting", "archive")).toThrow();
    });

    test("cannot archive from approved", () => {
      // approved can only go to applied, not archived
      expect(() => cr.apply("approved", "archive")).toThrow();
    });

    test("cannot apply from rejected", () => {
      expect(() => cr.apply("rejected", "apply")).toThrow();
    });

    test("cannot submit from archived", () => {
      expect(() => cr.apply("archived", "submit")).toThrow();
    });
  });
});
